Update EIP-7069: Rename and update semantics (#8287)
ethereum/EIPs: EIPS/eip-7069.md

We introduce four new instructions:

- `EXTCALL` (`0xf8`) with arguments `(target_address, input_offset, input_size, value)` 
- `EXTDELEGATECALL` (`0xf9`) with arguments `(target_address, input_offset, input_size)` - delegate call

> - `EXTSTATICCALL` (`0xfb`) with arguments `(target_address, input_offset, input_size)` - static call

Reviewer: The `- static call` comment I think is pointless now?

- `RETURNDATALOAD` (`0xf7`) with argument `offset`

Execution semantics of `EXT*CALL`:

1. Charge `WARM_STORAGE_READ_COST` (100) gas.
2. Pop required arguments from stack, fail with error on stack underflow.
3. If `value` is non-zero:
    3a. Fail with error if the current frame is in `static-mode`.
    3b. Charge `CALL_VALUE_COST` gas.
4. Peform (and charge for) memory expansion using `[input_offset, input_size]`.
5. If `target_address` is not in the `warm_account_list`, charge `COLD_ACCOUNT_ACCESS - WARM_STORAGE_READ_COST` (2500) gas.
6. If `target_address` is not in the state and the call configuration would result in account creation, charge `ACCOUNT_CREATION_COST` (25000) gas.
    - The only such case in this EIP is if `value` is non-zero.
7. Calculate the gas available to callee as caller's remaining gas reduced by `max(floor(gas/64), MIN_RETAINED_GAS)`.
8. Fail with error if the gas available to callee at this point is less than `MIN_CALLEE_GAS`.
9. If `value` is non-zero, fail lightly if the balance of the current account is less than `value`.

> 10. Fail lightly if call stack depth equals `1024`.

Reviewer 2: What does "fail Lightly" mean?  dose "Fail with error" mean revert and take all available gas and "fail lightly" mean fail consuming only already charged gas?

Author: yes (except "Fail with error" is not a revert but an exceptional abort, I presume this was the original intention), I'll specify these.

But you've uncovered an interesting issue - what should be pushed to stack in this case? For a light failure in EOFCREATE/TXCREATE we return `0` (with a different meaning), same for CREATE/CREATE2, but for revamped calls that would mean success.

I also see that `evmone` would return `0`, so making a self note to amend this (I'm working on that now in parallel).

Author: Proposing a resolution (`3` on stack for light failure) https://github.com/ethereum/EIPs/pull/8287/commits/54f6dcd92c95a21841e68fd7888dab5e02f6f51a to push the discussion forward. WDYT?

Author: https://github.com/ethereum/EIPs/pull/8287/commits/477e269900f21a4cdd6f48dfe08f36a8a750af79 adheres to what we spoke about on the EOF implementers call - `1` as status code for light failure

> 8. Halt with exceptional failure if the gas available to callee at this point is less than `MIN_CALLEE_GAS`.

Reviewer 3: Why is this exceptional failure? Seems could be light failure

Author: The PR only changes the language on this one. And I see it as a sensible failure mode - consistent with an OOG. But light failure also makes sense on 2nd thought. @chfast @axic @shemnon any more opinions on this, any objections to changing to light failure?

> 9. If `value` is non-zero, fail lightly if the balance of the current account is less than `value` (push `1` on the stack, only consume gas charged until this point).

Reviewer 3: kinda redundant
```suggestion
9. Fail lightly if the balance of the current account is less than `value` (push `1` on the stack, only consume gas charged until this point).
```

Author: wait, is caller nonce ever implemented for calls? Looks like too much copied from the similar CREATE case, I'll leave the nonce mention out.

Reviewer 3: Ah right, nonce is only relevant for creates, sorry for confusion.

11. Perform the call with the available gas and configuration.

> 12. Push a status code on the stack:
>     12a. `0` if the call was successful.
>     12b. `1` if the call has reverted.
>     12c. `2` if the call has failed.

Reviewer 2: This doesn't render as indended.

```suggestion
12. Push a status code on the stack:
    0. `0` if the call was successful.
    1. `1` if the call has reverted.
    2. `2` if the call has failed.
```

> 10. Fail lightly if call stack depth equals `1024` (push `1` on the stack, only consume gas charged until this point).

Reviewer 3: ```suggestion
10. Fail with `1` error code returned on stack, if current call stack depth equals `1024`. Only gas charged until this point is consumed.
```
(just "fail lightly" is not very commonly known term)

We introduce four new instructions:

- `EXTCALL` (`0xf8`) with arguments `(target_address, input_offset, input_size, value)` 
- `EXTDELEGATECALL` (`0xf9`) with arguments `(target_address, input_offset, input_size)`
- `EXTSTATICCALL` (`0xfb`) with arguments `(target_address, input_offset, input_size)`
- `RETURNDATALOAD` (`0xf7`) with argument `offset`

Execution semantics of `EXT*CALL`:

1. Charge `WARM_STORAGE_READ_COST` (100) gas.

> 2. Pop required arguments from stack, halt with exceptional failure on stack underflow.

Reviewer 3: It seems at this point EIP tries to maintain compatibility with both legacy and EOF, so maybe
```suggestion
2. Pop required arguments from stack, halt with exceptional failure on stack underflow.
  - *Note: When implemented in EOF, stack underflow check is done during stack validation, and runtime check is omitted.* 
```

Author: The idea is for this EIP to be fully legacy-complete and EOF-independent. EOF (Megaspec) lists the differences required it to be EOF-complete (which is lack of runtime stack underflow check and banning new opcodes in legacy).

Is there any other piece which is EOF-only in this EIP?

(this is just a question to your comment, I have no objections towards adding this note for completeness, and I'll add it)

3. If `value` is non-zero:
    - Halt with exceptional failure if the current frame is in `static-mode`.
    - Charge `CALL_VALUE_COST` gas.
4. Peform (and charge for) memory expansion using `[input_offset, input_size]`.
5. If `target_address` is not in the `warm_account_list`, charge `COLD_ACCOUNT_ACCESS - WARM_STORAGE_READ_COST` (2500) gas.
6. If `target_address` is not in the state and the call configuration would result in account creation, charge `ACCOUNT_CREATION_COST` (25000) gas.
    - The only such case in this EIP is if `value` is non-zero.
7. Calculate the gas available to callee as caller's remaining gas reduced by `max(floor(gas/64), MIN_RETAINED_GAS)`.
8. Halt with exceptional failure if the gas available to callee at this point is less than `MIN_CALLEE_GAS`.
9. If `value` is non-zero, fail lightly if the balance of the current account is less than `value` (push `3` on the stack, only consume gas charged until this point).

> 10. Fail lightly if call stack depth equals `1024` (push `3` on the stack, only consume gas charged until this point).

Reviewer 3: Call depth check is actually superseded by 63/64 rule, see Rationale.

Author: I thought you questioned the correctness of that initial removal of call depth check.

Isn't it a compatibility problem, that call-stack-depth-checking calls can alternate with non-checking ones in the call stack?

Why wasn't the call depth check removed already in legacy calls, if superseded by the 63/64 rule?

Author: for future reference - decided on the EOF Implementers call to keep it
1. least change to legacy
2. being cautious about the air-tightness of the (complex) gas-based 63/64 rule, and providing a simpler hard cap
3. being open to gas limits which are huge for various use cases beyond mainnet

We have changed the ruleset:

1. Use the 63/64th rule, but
    - Ensure that at least `MIN_RETAINED_GAS` gas is retained prior to executing the callee,
    - Ensure that at least `MIN_CALLEE_GAS` gas is available to the callee.

> Removing the call stack depth check was initially considered, but this would be incompatible with the original `*CALL` instructions, as well as `CREATE*` instructions, which can be intertwined with the new `EXT*CALL` instructions in the call stack. As such, keepeing the call stack depth involves the least change for legacy code.

Reviewer 3: maybe
```suggestion
Removing the call stack depth check was initially considered, but this would be incompatible with the original `*CALL` instructions, as well as `CREATE*` instructions, which can be intertwined with the new `EXT*CALL` instructions in the call stack. As such, keeping the call stack depth involves no change affecting legacy code.
```

We have changed the ruleset:

> 1. Use the 63/64th rule, but
>     - Ensure that at least `MIN_RETAINED_GAS` gas is retained prior to executing the callee,
>     - Ensure that at least `MIN_CALLEE_GAS` gas is available to the callee.

Reviewer 3: to remove weird 1-item numbered list
```suggestion
The 63/64th rule is still applied, but
    - At least `MIN_RETAINED_GAS` gas is retained prior to executing the callee,
    - At least `MIN_CALLEE_GAS` gas is available to the callee.
```

Removing the call stack depth check was initially considered, but this would be incompatible with the original `*CALL` instructions, as well as `CREATE*` instructions, which can be intertwined with the new `EXT*CALL` instructions in the call stack. As such, keeping the call stack depth check involves no change affecting legacy code.

> Also, we find the simple (as opposed to the complex 63/64th rule) hard cap reassuring, that the call stack depth is limited, in case the gas rules can be bypassed. Lastly, the amount of gas to reach depth of 1024 is huge, but not absurdly huge and we want to avoid constrain ourselves by relying this check on current gas limits.

Reviewer 3: ```suggestion
Also, we find the simple (as opposed to the complex 63/64th rule) hard cap reassuring, that the call stack depth is limited, in case the gas rules can be bypassed. Lastly, the amount of gas to reach depth of 1024 is huge, but not absurdly huge, and we want to avoid constraining ourselves by dependency of this check on current gas limits.
```

We change the value from boolean to a status code, where `0` signals success and thus it will be possible to introduce more non-success codes in the future, if desired.

> Status code `1` is used for both reverts coming from the callee frame and light failures encountered in the execution of the instructions. The reason for combining them is keeping the semantics similar to the original CALLs - both scenario preserve unused gas and continue being indistinguishable to the caller.

Reviewer 3: ```suggestion
Status code `1` is used for both reverts coming from the callee frame and light failures encountered in the execution of the instructions. The reason for combining them is keeping the semantics similar to the original CALLs - both scenarios preserve unused gas and continue being indistinguishable to the caller.
```
